autosklearn/ensemble_builder.py

```diff
@@ -691,25 +697,81 @@ def predict(self, set_: str,
         # TODO: ADD saving of predictions on "ensemble data"
 
     def _delete_non_candidate_models(self, candidates):
-        candidates = [os.path.split(cand)[1] for cand in candidates]
-        for model_path in self.read_preds.keys():
-            if self.read_preds[model_path]['deleted']:
+        # Loop through the files currently in the directory
+        for pred_path in self.y_ens_files:
+
+            # Do not delete candidates
+            if pred_path in candidates:
                 continue
-            match = self.model_fn_re.search(model_path)
-            _num_run = int(match.group(2))
-            # Do not remove the dummy prediction!
-            if _num_run == 1:
+
+            match = self.model_fn_re.search(pred_path)
+            _full_name = match.group(0)
+            _seed = match.group(1)
+            _num_run = match.group(2)
+            _budget = match.group(3)
+
+            # Do not delete the dummy prediction
+            if int(_num_run) == 1:
                 continue
-            model_file = os.path.split(model_path)[1]
-            if model_file not in candidates:
+
+            # Besides the prediction, we have to take care of three other files: model,
+            # validation and test.
+            model_name = '%s.%s.%s.model' % (_seed, _num_run, _budget)
+            model_path = os.path.join(self.dir_models, model_name)
+            pred_valid_name = 'predictions_valid' + _full_name
+            pred_valid_path = os.path.join(self.dir_valid, pred_valid_name)
+            pred_test_name = 'predictions_test' + _full_name
+            pred_test_path = os.path.join(self.dir_test, pred_test_name)
+
+            # Lets lock all the files "at once" to avoid weird race conditions. This is
```

Comment:
I can see why this most likely won't fail, but I'm wondering whether we should make this really failsafe. We could approach this by "locking all files at once", i.e. placing a lock on the validation file and only if we acquire that delete all related files. Then, we'd also have to slightly change the logic in `autosklearn.evaluation.abstract_evaluator.AbstractEvaluator.file_output` to acquire a full lock. In this way, we would either write or delete all files in an atomic manner.

Reply:
Agreed.
The full lock for deleting is already what this code does. If we implement the full lock for writing than we are failsafe. I'm going to have a go at that.